app/preload/accountReport.ts

```diff
@@ -0,0 +1,274 @@
+    isOutgoing: boolean
+): string {
+    if (
+        [TransactionStatus.Failed, TransactionStatus.Rejected].includes(
+            transaction.status
+        )
+    ) {
+        return '0';
+    }
+    if (isShieldedBalanceTransaction(transaction)) {
+        if (!transaction.decryptedAmount) {
+            return '?';
+        }
+        switch (transaction.transactionKind) {
+            case TransactionKindString.TransferToPublic:
+            case TransactionKindString.TransferToEncrypted:
+                return transaction.decryptedAmount;
+            case TransactionKindString.EncryptedAmountTransfer:
+            case TransactionKindString.EncryptedAmountTransferWithMemo:
+                return isOutgoing
+                    ? '-'.concat(transaction.decryptedAmount)
+                    : transaction.decryptedAmount;
+            default:
+                throw new Error('unexpected transaction type');
+        }
+    } else {
+        return '0';
+    }
+}
+
+const getName = (i: string[]) => i[0];
+const getLabel = (i: string[]) => i[1];
+const exportedFields = Object.entries(exportTransactionFields);
+
+// Parse a transaction into a array of values, corresponding to those of the exported fields.
+function parseTransaction(
+    transaction: TransferTransactionWithNames,
+    address: string
+) {
+    const fieldValues: Record<string, string> = {};
+    Object.entries(transaction).forEach(([key, value]) => {
+        fieldValues[key] = value?.toString();
+    });
+
+    fieldValues.dateTime = getISOFormat(transaction.blockTime);
+    fieldValues.transactionKind =
+        transactionKindNames[transaction.transactionKind];
+
+    const isOutgoing = transaction.fromAddress === address;
+    fieldValues.publicBalance = calculatePublicBalanceChange(
+        transaction,
+        isOutgoing
+    );
+    fieldValues.shieldedBalance = calculateShieldedBalanceChange(
+        transaction,
+        isOutgoing
+    );
+    if (!isOutgoing) {
+        fieldValues.cost = '';
+    }
+
+    return exportedFields.map((field) => fieldValues[getName(field)]);
+}
+
+/** Given a nested list of strings, parses the elements to csv,
+ *   each inner list is turned into a line.
+ */
+function toCSV(elements: string[][]): string {
+    return elements.map((element) => element.join(',')).join('\n');
+}
+
+/**
+ * Streams the given account's transactions into the given writable, csv formatted.
+ * @param stream a Writable, which the transactions are written to.
+ * @param account account from which transactions are selected.
+ */
+async function streamTransactions(
+    stream: Writable,
+    account: Account,
+    filter: TransactionFilter,
+    abortController: AbortController
+) {
+    const getAddressName = async (address: string) => {
+        if (!address) {
+            return undefined;
+        }
+        if (address === account.address) {
+            return account.name;
+        }
+        return getEntryName(address);
```

Comment:
A potential optimization (that we probably don't need right now) would be to cache the names we look up here, as if we see an address once there is probably a high chance of seeing it again.

Reply:
Yea, I did the obvious one, with the account, which are making the report.